evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,372 @@
+                parent = parent[:-2]
+            ancestors.append(parent)
+            parents.extend(self.get_parents(parent))
+        ancestors.reverse()
+        return ancestors
+
+    def _compute_features(self, component_list, X, y=None, fit=False):
+        """Transforms the data by applying the given components.
+
+        Arguments:
+            component_list (list): The list of component names to compute.
+            X (pd.DataFrame): Input data to the pipeline to transform.
+            y (pd.Series): The target training data of length [n_samples]
+            fit (bool): Whether to fit the estimators as well as transform it.
+                        Defaults to False.
+
+        Returns:
+            pd.DataFrame or pd.Series - Output of the last component
+        """
+        if not isinstance(X, pd.DataFrame):
+            X = pd.DataFrame(X)
+
+        output_cache = {}
+        final_component = None
+        for component_name in component_list:
+            final_component = component_name
+            component_class = self.get_component(component_name)
+            if not isinstance(component_class, ComponentBase):
+                raise ValueError('All components must be instantiated before fitting or predicting')
+            x_inputs = []
+            y_input = None
+            for parent_input in self.get_parents(component_name):
+                if parent_input[-2:] == '.y':
+                    if y_input is not None:
+                        raise ValueError(f'Cannot have multiple `y` parents for a single component {component_name}')
+                    y_input = output_cache[parent_input]
+                else:
+                    try:
+                        parent_x = output_cache[parent_input]
+                        if isinstance(parent_x, pd.Series):
+                            parent_x = pd.DataFrame(parent_x, columns=[parent_input])
+                        x_inputs.append(parent_x)
+                    except KeyError:
+                        x_inputs.append(output_cache[f'{parent_input}.x'])
+            input_x, input_y = self._consolidate_inputs(x_inputs, y_input, X, y)
+            self.input_feature_names.update({component_name: list(input_x.columns)})
+            if isinstance(component_class, Transformer):
+                if fit:
+                    output = component_class.fit_transform(input_x, input_y)
+                else:
+                    output = component_class.transform(input_x, input_y)
+                if isinstance(output, tuple):
+                    output_x, output_y = output[0], output[1]
+                else:
+                    output_x = output
+                    output_y = None
+                output_cache[f"{component_name}.x"] = output_x
+                output_cache[f"{component_name}.y"] = output_y
+            else:
+                if fit:
+                    component_class = component_class.fit(input_x, input_y)
```

Comment:
Here you can just say
```python
component_class.fit(input_x, input_y)
```
No need to save the return value since you already have it :)

Reply:
This is a good idea, but when I made this change a couple of tests fail because the mocked `Estimator.predict` is getting called one fewer time. Do you have any idea why that might be?

Reply:
@eccabay oh that's quite strange. Do you happen to remember which tests were failing -- was it [`test_fit` and similar](https://github.com/alteryx/evalml/pull/1415/files#diff-b878b993a49832dd7425f3daea137f01a41f799d0333851e618e6244236f3d4aR340)? The behavior of our components shouldn't change based on whether or not we save the output of fit, since [fit just returns `self`](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/components/component_base.py#L85). Its possible one of the estimators in the example graph in that test violates that somehow, but I'm not sure how.

If you're able to reproduce that behavior outside of the context of this code, I think we should file it as a bug.

Reply:
Actually after looking into this, it was partly just me being silly. The mocked `Estimator.predict` _should_ have been being called one less time than I thought, since I put in a catch not to call predict on the final estimator during fit. Somehow, setting `component_class = component_class.fit()` messed with that logic and was preventing the catch from working properly. 

TL,DR: I've deleted this and all is well